code/globalincs/pool.h

```diff
@@ -0,0 +1,404 @@
+				entry.stored.reset();
+		}
+		for (size_t i=storage.size(); i<cap;i++){
+			gEntry<T> e;
+			e.generation = 0;
+			e.stored = tl::nullopt;
+			storage.push_back(e);
+		}
+		fill_empty_list();
+	}
+
+	//Index into the storage... bare reference
+	T& operator[](pool_index i) {
+		if (i.null())
+			throw PoolExceptionNullIndex("null index SCP_Pool access attempted");
+		if (i.i() >= storage.size())
+			throw PoolExceptionOutOfRange("out of range SCP_Pool access attempted");
+		if (storage[i.i()].generation!= i.g())
+			throw PoolExceptionStaleIndex("Stale SCP_Pool access attempted");
+		if(storage[i.i()].stored==tl::nullopt)
+			throw PoolExceptionNullValue("Empty SCP_Pool access attempted");
+//		v = (storage[i.i()].stored).operator->();
+		return storage[i.i()].stored.operator*();
+	};
+
+	optional<T&> get(pool_index i){
+		SCP_Pool<T> &self = this;
+		try{
+			T& r = self[i];
+			return optional<T&>(r);
+		}
+		catch(const PoolExceptionNullIndex& e) {
+			return tl::nullopt;
+		}
+		catch(const PoolExceptionOutOfRange& e) {
+			return tl::nullopt;
+		}
+		catch(const PoolExceptionStaleIndex& e) {
+			return tl::nullopt;
+		}
+		catch(const PoolExceptionNullValue& e) {
+			return tl::nullopt;
+		}
+	};
+	optional<T*> get_pointer(pool_index i){
+		if (i.i() >= storage.size()) {
+			return tl::nullopt;
+		}
+		if (storage[i.i()].generation!= i.g()){
+			return tl::nullopt;
+		}
+		optional<T> s = storage[i.i()].stored;
+		T* p = s.operator->();
+		return p;
+	};
+
+	optional<pool_index> get_new(){
+		T* n = new T();
+		return store(*n);
+	};
+
+	//Store a value in the vector
+	//TODO: Handle capped stuff
+	//This is kind of in appropriate for the actual usage this is primarily intended to replace
+	//to the point where if we want this functionality it might need to be a seperate data structure for sake of strictness
+	optional<pool_index> store(T &n){
+		if (known_empty.empty()) {
+			if (capped && storage.size()>=cap) {
+				return tl::nullopt;
+			}
+			return add_new(n);
+		}
+		return use_free(n);
+	};
+
+	void remove(optional<pool_index> i){
+		if (!i.has_value())
+			return;
+		remove(i.value());
+	};
+	void remove(pool_index i){
+		if (i.null())
+			throw PoolExceptionNullIndex("Attempted to remove from SCP_Pool with a null index");
+		if (i.i() >= storage.size())
+			throw PoolExceptionOutOfRange("Attempted to remove from SCP_Pool with out-of-range index");
+		if (storage[i.i()].generation != i.g() ) 
+			throw PoolExceptionStaleIndex("Attempted to remove from SCP_Pool with stale index");
+		reset_slot(i.i());
+	};
+	bool check(pool_index i){
+		if(i.null())
+			return false;
+		auto * t = &operator[](i);
+		return (t!=nullptr);
+	}
+
+
+
+	//iterator shit.
+	struct iter{
+		using iterator_category = std::input_iterator_tag;
+		using index = int;
+		using storage_t = vector<gEntry<T>>;
+		using storage_p = vector<gEntry<T>>*;
+
+		iter(SCP_Pool<T> *vec_ptr, storage_p storage_ptr){
+			storage = storage_ptr;
+			vec = vec_ptr;
+		};
```

Comment:
This construtor probably should not exist, or at the very least, just be an alias to iter(..., ..., 0), as it allows to create an iterator that points to an invalid element if element 0 is empty